compiler/rustc_resolve/src/lib.rs

```diff
@@ -1132,8 +1132,8 @@ impl ResolverAstLowering for Resolver<'_> {
         self.next_node_id()
     }
 
-    fn trait_map(&self) -> &NodeMap<Vec<TraitCandidate>> {
-        &self.trait_map
+    fn trait_map(&mut self) -> NodeMap<Vec<TraitCandidate>> {
+        std::mem::take(&mut self.trait_map)
```

Comment:
Any subsequent calls to `trait_map()` will return an empty `NodeMap`, which seems like a footgun. Could you add a (debug) assertion to verify that this is only called once?